src/mscorlib/src/System/String.Comparison.cs

```diff
@@ -817,153 +817,106 @@ public override bool Equals(Object obj)
+        }
 
-            if ((Object)this == (Object)value) {
-                return true;
-            }
+        [Pure]
+        public static bool Equals(String a, String b, StringComparison comparisonType) {
+            return EqualsHelperComparison(a, b, comparisonType);
+        }
 
-            if ((Object)value == null) {
-                return false;
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool EqualsHelperComparison(String a, String b, StringComparison comparisonType)
+        {
+            if (comparisonType == StringComparison.Ordinal) { 
+                return EqualsOrdinalCaseSensitive(a, b);
             }
+            else if (comparisonType == StringComparison.OrdinalIgnoreCase) {
+                return EqualsOrdinalIgnoreCase(a, b);
+            }
+            else {
+                return EqualsCulture(a, b, comparisonType);
+            }
+        }
 
-            switch (comparisonType) {
-                case StringComparison.CurrentCulture:
-                    return (CultureInfo.CurrentCulture.CompareInfo.Compare(this, value, CompareOptions.None) == 0);
-
-                case StringComparison.CurrentCultureIgnoreCase:
-                    return (CultureInfo.CurrentCulture.CompareInfo.Compare(this, value, CompareOptions.IgnoreCase) == 0);
-
-                case StringComparison.InvariantCulture:
-                    return (CultureInfo.InvariantCulture.CompareInfo.Compare(this, value, CompareOptions.None) == 0);
-
-                case StringComparison.InvariantCultureIgnoreCase:
-                    return (CultureInfo.InvariantCulture.CompareInfo.Compare(this, value, CompareOptions.IgnoreCase) == 0);
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool EqualsOrdinalCaseSensitive(String a, String b) {
+            if (ReferenceEquals(a, b)) return true;
+            if (NotPossiblyEquals(a, b)) return false;
 
-                case StringComparison.Ordinal:
-                    if (this.Length != value.Length)
-                        return false;
-                    return EqualsHelper(this, value);
+            return EqualsHelper(a, b);
+        }
 
-                case StringComparison.OrdinalIgnoreCase:
-                    if (this.Length != value.Length)
-                        return false;
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool EqualsOrdinalIgnoreCase(String a, String b) {
+            if (ReferenceEquals(a, b)) return true;
+            if (NotPossiblyEquals(a, b)) return false;
 
-                    // If both strings are ASCII strings, we can take the fast path.
-                    if (this.IsAscii() && value.IsAscii()) {
-                        return (CompareOrdinalIgnoreCaseHelper(this, value) == 0);
-                    }
+            var isAscii = a.IsAscii();
+            if (isAscii != b.IsAscii()) return false;
```

Comment:
Perhaps line 387 can also have this trick of returning false if exactly one is ascii.